src/Workspaces/Core/Portable/Workspace/Solution/SourceGeneratedDocumentState.cs

```diff
@@ -16,34 +16,75 @@ internal sealed class SourceGeneratedDocumentState : DocumentState
         public ISourceGenerator SourceGenerator { get; }
 
         public static SourceGeneratedDocumentState Create(
-            GeneratedSourceResult generatedSourceResult,
+            SourceGeneratedDocumentState? previousState,
+            string hintName,
+            SourceText generatedSourceText,
+            SyntaxTree generatedSyntaxTree,
             DocumentId documentId,
             ISourceGenerator sourceGenerator,
             HostLanguageServices languageServices,
-            SolutionServices solutionServices)
+            SolutionServices solutionServices,
+            CancellationToken cancellationToken)
         {
-            var textAndVersion = TextAndVersion.Create(generatedSourceResult.SourceText, VersionStamp.Create());
-            ValueSource<TextAndVersion> textSource = new ConstantValueSource<TextAndVersion>(textAndVersion);
+            var tree = generatedSyntaxTree;
+            var options = tree.Options;
+            var filePath = tree.FilePath;
+
+            bool sourceChanged;
+            ValueSource<TextAndVersion> textSource;
 
-            var tree = generatedSourceResult.SyntaxTree;
+            // We can reuse the previous TextAndVersion if the checksum from the text matches.
+            if (previousState is null
+                || !previousState.TryGetTextAndVersion(out var textAndVersion)
+                || Checksum.From(textAndVersion.Text.GetChecksum()) != Checksum.From(generatedSourceText.GetChecksum()))
+            {
+                sourceChanged = true;
+                textAndVersion = TextAndVersion.Create(generatedSourceText, VersionStamp.Create());
+                textSource = new ConstantValueSource<TextAndVersion>(textAndVersion);
+            }
+            else
+            {
+                sourceChanged = false;
+                textSource = previousState.TextAndVersionSource;
+            }
 
-            // Since the tree is coming directly from the generator, this tree is strongly held so GetRoot() doesn't need a CancellationToken.
-            var root = tree.GetRoot(CancellationToken.None);
-            Contract.ThrowIfNull(languageServices.SyntaxTreeFactory, "We should not have a generated syntax tree for a language that doesn't support trees.");
+            // We can reuse the previous tree if the options haven't changed
+            bool treeChanged;
+            if (!sourceChanged && Equals(previousState!.ParseOptions, generatedSyntaxTree.Options))
+            {
+                treeChanged = false;
+                tree = previousState.SyntaxTree.WithFilePath(tree.FilePath);
```

Comment:
why the .WithFilePath here?  can you doc.  it's not apparent to me why this is needed.

Reply:
Doubly so after assigning treeChanged to false!